// gatsby-cli/lib/reporter/loggers/ink/components/develop.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _ink = require("ink");

var _context = _interopRequireDefault(require("../context"));

// Track the width and height of the terminal. Responsive app design baby!
const useTerminalResize = () => {
  const _useContext = (0, _react.useContext)(_ink.StdoutContext),
        stdout = _useContext.stdout;

  const _useState = (0, _react.useState)([stdout.columns, stdout.rows]),
        sizes = _useState[0],
        setSizes = _useState[1];

  (0, _react.useEffect)(() => {
    const resizeListener = () => {
      setSizes([stdout.columns, stdout.rows]);
    };

    stdout.on(`resize`, resizeListener);
    return () => {
      stdout.off(`resize`, resizeListener);
    };
  }, [stdout]);
  return sizes;
};

const Develop = ({
  pagesCount,
  appName,
  status
}) => {
  const _useTerminalResize = useTerminalResize(),
        width = _useTerminalResize[0];

  return _react.default.createElement(_ink.Box, {
    flexDirection: "column",
    marginTop: 2
  }, _react.default.createElement(_ink.Box, {
    textWrap: `truncate`
  }, `—`.repeat(width)), _react.default.createElement(_ink.Box, {
    height: 1,
    flexDirection: "row"
  }, _react.default.createElement(_ink.Color, null, pagesCount, " pages"), _react.default.createElement(_ink.Box, {
    flexGrow: 1
  }), _react.default.createElement(_ink.Color, null, status), _react.default.createElement(_ink.Box, {
    flexGrow: 1
  }), _react.default.createElement(_ink.Color, null, appName)));
};

const ConnectedDevelop = () => {
  var _state$pages, _state$program, _state$logs;

  const state = (0, _react.useContext)(_context.default);
  return _react.default.createElement(Develop, {
    pagesCount: ((_state$pages = state.pages) === null || _state$pages === void 0 ? void 0 : _state$pages.size) || 0,
    appName: ((_state$program = state.program) === null || _state$program === void 0 ? void 0 : _state$program.sitePackageJson.name) || ``,
    status: ((_state$logs = state.logs) === null || _state$logs === void 0 ? void 0 : _state$logs.status) || ``
  });
};

var _default = ConnectedDevelop;
exports.default = _default;
